blockly-ws-search/src/WorkspaceSearch.js

```diff
@@ -0,0 +1,660 @@
+   *    if it is included in the new matching blocks.
+   */
+  searchAndHighlight(preserveCurrent) {
+    let oldCurrentBlock = this.currentBlock_;
+    this.clearBlocks();
+    this.blocks_ = this.getMatchingBlocks_(
+        this.workspace_, this.searchText_, this.caseSensitive);
+    this.highlightSearchGroup(this.blocks_);
+    let currentIdx = 0;
+    if (preserveCurrent) {
+      currentIdx = this.blocks_.indexOf(oldCurrentBlock);
+      currentIdx = currentIdx > -1 ? currentIdx : 0;
+    }
+    this.setCurrentBlock_(currentIdx);
+  }
+
+  /**
+   * Returns pool of blocks to search from.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to get blocks from.
+   * @return {!Array.<!Blockly.BlockSvg>} The search pool of blocks to use.
+   * @private
+  */
+  getSearchPool_(workspace) {
+    const blocks = (
+        /** @type {!Array.<!Blockly.BlockSvg>} */
+        workspace.getAllBlocks(true));
+    return blocks.filter(function(block) {
+      // Filter out blocks contained inside of another collapsed block.
+      const surroundParent = block.getSurroundParent();
+      return !surroundParent || !surroundParent.isCollapsed();
+    });
+  }
+
+  /**
+   * Returns whether the given block matches the search text.
+   * @param {!Blockly.BlockSvg} block The block to check.
+   * @param {string} searchText The search text. Note if the search is case
+   *    insensitive, this will be passed already converted to lowercase letters.
+   * @param {boolean} caseSensitive Whether the search is caseSensitive.
+   * @return {boolean} Whether the block matches the search text.
+   * @private
+   */
+  isBlockMatch_(block, searchText, caseSensitive) {
+    let blockText = '';
+    if (block.isCollapsed()) {
+      // Search the whole string for collapsed blocks.
+      blockText = block.toString();
+    } else {
+      const topBlockText = [];
+      block.inputList.forEach(function(input) {
+        input.fieldRow.forEach(function(field) {
+          topBlockText.push(field.getText());
+        });
+      });
+      blockText = topBlockText.join(' ').trim();
+    }
+    if (!caseSensitive) {
+      blockText = blockText.toLowerCase();
+    }
+    return blockText.includes(searchText);
+  }
+
+  /**
+   * Returns blocks that match the given search text.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to search.
+   * @param {string} searchText The search text.
+   * @param {boolean} caseSensitive Whether the search should be case sensitive.
+   * @return {!Array.<Blockly.BlockSvg>} blocks The blocks that match the search
+   *    text.
+   * @protected
+   */
+  getMatchingBlocks_(workspace, searchText, caseSensitive) {
+    if (!searchText) {
+      return [];
+    }
+    if (!this.caseSensitive) {
+      searchText = searchText.toLowerCase();
+    }
+    const searchGroup = this.getSearchPool_(workspace);
+    return searchGroup.filter(
+        block => this.isBlockMatch_(block, searchText, caseSensitive));
```

Comment:
What is the purpose of splitting this method into another method. Do we expect to use isBlockMatch_ elsewhere? This shallow splitting of methods is actually making it harder to follow the code IMO.

Reply:
I split it up because I thought it was easier to read.

Reply:
Okay this is going to be hard to reason about as its very subjective, but since both of you think it reads better keep as is.